manticore/core/cpu/abstractcpu.py

```diff
@@ -591,9 +637,7 @@ def read_string(self, where, max_length=None):
         while True:
             c = self.read_int(where, 8)
 
-            assert not issymbolic(c)
-
-            if c == 0:
+            if c == 0 or issymbolic(c):
```

Comment:
we should check `issymbolic` first; otherwise symbolic needs to implement `__nonzero__`

```
2017-11-28 17:23:25,775: [37608] m.c.executor:ERROR: Exception: __nonzero__ for Bool
Traceback (most recent call last):
  File "/mnt/hgfs/code/manticore/manticore/core/executor.py", line 460, in run
    if not current_state.execute():
  File "/mnt/hgfs/code/manticore/manticore/core/state.py", line 126, in execute
    result = self._platform.execute()
  File "/mnt/hgfs/code/manticore/manticore/platforms/linux.py", line 2010, in execute
    self.syscall()
  File "/mnt/hgfs/code/manticore/manticore/platforms/linux.py", line 1855, in syscall
    return self._syscall_abi.invoke(implementation)
  File "/mnt/hgfs/code/manticore/manticore/core/cpu/abstractcpu.py", line 390, in invoke
    s = self._cpu.read_string(arg, max_arg_expansion).translate(None, '\n')
  File "/mnt/hgfs/code/manticore/manticore/core/cpu/abstractcpu.py", line 640, in read_string
    if c == 0 or issymbolic(c):
  File "/mnt/hgfs/code/manticore/manticore/core/smtlib/expression.py", line 119, in __nonzero__
    raise NotImplementedError("__nonzero__ for Bool")
NotImplementedError: __nonzero__ for Bool

```

Reply:
To implement __nonzero__ from Expression you need to reference current _constraints_ whatever that is from the expression.
Also we probably do not want that as __nonzero__ normally should fork. 
We can though improve the __nonzero__ not implemented exception.   

((( Or the craziness catch the __nonzero__ exception and replace the value in the python call stack with one of the possible concretizations of the exp )))

Reply:
+1, think checking for `issymbolic` first is the right way to go here.